File: spec/lib/api/v1/outcome_spec.rb
#
# Copyright (C) 2011 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

require File.expand_path(File.dirname(__FILE__) + '/../../../spec_helper.rb')

lib = (Class.new do
  include Api::V1::Outcome

  def api_v1_outcome_path(opts)
    "/api/v1/outcome/#{opts.fetch(:id)}"
  end
s
  def polymorphic_path(*_args)
    '/test'
  end
end).new

RSpec.describe "Api::V1::Outcome" do
  def new_outcome(creation_params = {})
    creation_params.reverse_merge!({
      :title => 'TMNT Beats',
      :calculation_method => 'decaying_average',
      :calculation_int => 65,
      :display_name => "Ninja Rap",
      :description => "Turtles with Vanilla Ice",
      :vendor_guid => "TurtleTime4002",
    })

    @outcome = LearningOutcome.create!(creation_params)
    @outcome
  end

  def new_outcome_link(creation_params = {}, course = nil)
    outcome = new_outcome(creation_params) # sets @outcome
    unless course
      course_with_teacher(active_all: true)  # sets @course
      course = @course
    end
    course.root_outcome_group.add_outcome(outcome)
  end

  context "json" do
    let(:outcome_params) do
      {
        :title => 'TMNT Beats',
        :calculation_method => 'decaying_average',
        :calculation_int => 65,
        :display_name => "Ninja Rap",
        :description => "Turtles with Vanilla Ice",
        :vendor_guid => "TurtleTime4002",
      }
    end

    context "outcome json" do
      let(:check_outcome_json) do
        ->(outcome) do
          expect(outcome['title']).to eq(outcome_params[:title])
          expect(outcome['calculation_method']).to eq(outcome_params[:calculation_method])
          expect(outcome['calculation_int']).to eq(outcome_params[:calculation_int])
          expect(outcome['display_name']).to eq(outcome_params[:display_name])
          expect(outcome['description']).to eq(outcome_params[:description])
          expect(outcome['vendor_guid']).to eq(outcome_params[:vendor_guid])
          expect(outcome['assessed']).to eq(LearningOutcome.find(outcome['id']).assessed? ? true : false)
          expect(outcome['has_updateable_rubrics']).to eq(
            LearningOutcome.find(outcome['id']).updateable_rubrics?
          )
          expect(outcome['ratings'].length).to eq 3
        end
      end

      it "returns the json for an outcome" do
        check_outcome_json.call(lib.outcome_json(new_outcome(outcome_params), nil, nil))
      end

      it "returns the json for multiple outcomes" do
        outcomes = []
        10.times{ outcomes.push(new_outcome) }
        lib.outcomes_json(outcomes, nil, nil).each { |o| check_outcome_json.call(o) }
      end
    end

    context "outcome links json" do
      let(:check_outcome_link_json) do
        ->(outcome, course, outcome_link) do
          expect(outcome_link['outcome']['id']).to eq(outcome.id)
          expect(outcome_link['outcome']['title']).to eq(outcome_params[:title])
          expect(outcome_link['outcome']['vendor_guid']).to eq(outcome_params[:vendor_guid])

          expect(outcome_link['context_type']).to eq("Course")
          expect(outcome_link['context_id']).to eq(course.id)
          expect(outcome_link['url']).to eq(lib.polymorphic_path)

          expect(outcome_link['outcome_group']['id']).to eq(course.root_outcome_group.id)
        end
      end

      it "returns the json for an outcome link" do
        outcome_link = new_outcome_link(outcome_params)
        check_outcome_link_json.call(@outcome, @course, lib.outcome_link_json(outcome_link, nil, nil))
      end

      it "returns the json for multiple outcome links" do
        course_with_teacher(active_all: true)  # sets @course
        outcome_links = 10.times.map{ new_outcome_link(outcome_params, @course) }
        lib.outcome_links_json(outcome_links, nil, nil).each do |ol|
          check_outcome_link_json.call(
            LearningOutcome.find(ol["outcome"]["id"]),
            @course,
            ol
          )
        end
      end
    end
  end
end

File: spec/coffeescripts/models/OutcomeSpec.js
/*
 * Copyright (C) 2015 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import Outcome from '@canvas/outcomes/backbone/models/Outcome'
import fakeENV from 'helpers/fakeENV'

QUnit.module('Course Outcomes', {
  setup() {
    this.importedOutcome = {
      context_type: 'Account',
      context_id: 1,
      outcome: {
        title: 'Account Outcome',
        context_type: 'Course',
        context_id: 1,
        calculation_method: 'decaying_average',
        calculation_int: 65,
      },
    }
    this.courseOutcome = {
      context_type: 'Course',
      context_id: 2,
      outcome: {
        title: 'Course Outcome',
        context_type: 'Course',
        context_id: 2,
      },
    }
    fakeENV.setup()
    ENV.PERMISSIONS = {manage_outcomes: true}
    ENV.ROOT_OUTCOME_GROUP = {context_type: 'Course'}
  },
  teardown() {
    fakeENV.teardown()
  },
})

test('isNative returns false for an outcome imported from the account level', function () {
  const outcome = new Outcome(this.importedOutcome, {parse: true})
  equal(outcome.isNative(), false)
})

test('isNative returns true for an outcome created in the course', function () {
  const outcome = new Outcome(this.courseOutcome, {parse: true})
  equal(outcome.isNative(), true)
})

test('CanManage returns true for an account outcome on the course level', function () {
  const outcome = new Outcome(this.importedOutcome, {parse: true})
  equal(outcome.canManage(), true)
})

test('default calculation method settings not set if calculation_method exists', function () {
  const spy = sandbox.spy(Outcome.prototype, 'setDefaultCalcSettings')
  new Outcome(this.importedOutcome, {parse: true})
  ok(!spy.called)
})

test('default calculation method settings set if calculation_method is null', function () {
  const spy = sandbox.spy(Outcome.prototype, 'setDefaultCalcSettings')
  new Outcome(this.courseOutcome, {parse: true})
  ok(spy.calledOnce)
})

QUnit.module('Account Outcomes', {
  setup() {
    this.accountOutcome = {
      context_type: 'Account',
      context_id: 1,
      outcome: {
        title: 'Account Outcome',
        context_type: 'Account',
        context_id: 1,
      },
    }
    fakeENV.setup()
    ENV.PERMISSIONS = {manage_outcomes: true}
    ENV.ROOT_OUTCOME_GROUP = {context_type: 'Account'}
  },
  teardown() {
    fakeENV.teardown()
  },
})

test('isNative is true for an account level outcome when viewed on the account', function () {
  const outcome = new Outcome(this.accountOutcome, {parse: true})
  equal(outcome.isNative(), true)
})

QUnit.module('Global Outcomes in a course', {
  setup() {
    this.globalOutcome = {
      context_type: 'Account',
      context_id: 1,
      outcome: {
        title: 'Account Outcome',
        context_type: undefined,
        context_id: undefined,
      },
    }
    fakeENV.setup()
    ENV.PERMISSIONS = {manage_outcomes: true}
    ENV.ROOT_OUTCOME_GROUP = {context_type: 'Course'}
  },
  teardown() {
    fakeENV.teardown()
  },
})

test('CanManage returns true for a global outcome on the course level', function () {
  const outcome = new Outcome(this.globalOutcome, {parse: true})
  equal(outcome.canManage(), true)
})

QUnit.module('With the account_level_mastery_scales FF enabled', {
  setup() {
    this.importedOutcome = {
      context_type: 'Account',
      context_id: 1,
      outcome: {
        title: 'Account Outcome',
        context_type: 'Course',
        context_id: 1,
        calculation_method: 'decaying_average',
        calculation_int: 65,
      },
    }
    fakeENV.setup()
    this.ratings = [
      {description: 'Exceeds Mastery', points: 4.0, mastery: false, color: '127A1B'},
      {description: 'Mastery', points: 3.0, mastery: true, color: '0B874B'},
      {description: 'Near Mastery', points: 2.0, mastery: false, color: 'FAB901'},
      {description: 'Below Mastery', points: 1.0, mastery: false, color: 'FD5D10'},
      {description: 'Well Below Mastery', points: 0.0, mastery: false, color: 'E0061F'},
    ]
    ENV.PERMISSIONS = {manage_outcomes: true}
    ENV.ROOT_OUTCOME_GROUP = {context_type: 'Course'}
    ENV.MASTERY_SCALE = {
      outcome_proficiency: {ratings: this.ratings},
    }
    ENV.ACCOUNT_LEVEL_MASTERY_SCALES = true
  },
  teardown() {
    fakeENV.teardown()
  },
})

test('it uses the ENV.MASTERY_SCALES ratings', function () {
  const outcome = new Outcome(this.importedOutcome, {parse: true})
  equal(outcome.get('ratings'), this.ratings)
  equal(outcome.get('mastery_points'), 3)
  equal(outcome.get('points_possible'), 4)
})

test('ignores proficiency attributes during saving', function () {
  const outcome = new Outcome(this.importedOutcome, {parse: true})
  sinon.stub(outcome, 'url').returns('fake-url')
  outcome.save({}, {})
  equal(outcome.get('mastery_points'), null)
  equal(outcome.get('points_possible'), null)
  equal(outcome.get('ratings'), null)
  equal(outcome.get('calculation_method'), null)
  equal(outcome.get('calculation_int'), null)
})
